File: backend/app/schemas/flight.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator

RankPreference = Literal["cheap", "fast", "balanced"]
CabinClass = Literal["economy", "premium_economy", "business", "first"]


class FlightSearchRequest(BaseModel):
    origin: str = Field(..., min_length=1, max_length=32, description="IATA / 中文 / Trip.com 城市码")
    destination: str = Field(..., min_length=1, max_length=32)
    date: str = Field(..., pattern=r"^\d{4}-\d{2}-\d{2}$")
    return_date: str | None = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    adults: int = Field(default=1, ge=1, le=9)
    cabin: CabinClass = "economy"
    preference: RankPreference = "balanced"
    include_letsfg: bool = False
    include_ignav: bool = True

    @field_validator("date", "return_date")
    @classmethod
    def validate_real_date(cls, v: str | None) -> str | None:
        if v is None:
            return v
        from datetime import date

        date.fromisoformat(v)
        return v


class PurchaseChannel(BaseModel):
    name: Literal["Trip.com"] = "Trip.com"
    url: str
    type: Literal["search"] = "search"
    note: str = "航班与价格以 Trip.com 预订页为准"


class FlightLegSnapshot(BaseModel):
    """单趟行程（含转机 segment 聚合后的门到门信息）。"""

    origin_iata: str
    dest_iata: str
    airline: str
    route_label: str
    depart_time: str
    arrive_time: str
    duration_minutes: int
    stops: int
    flight_numbers: list[str] = Field(default_factory=list)


class FlightQuote(BaseModel):
    """Ignav / LetsFG 等参考价源。flat 字段为去程/单程；往返时另含 return_leg。"""

    id: str
    origin_iata: str
    dest_iata: str
    airline: str
    route_label: str
    depart_time: str
    arrive_time: str
    duration_minutes: int
    stops: int
    price_amount: float
    price_currency: str
    source: str
    confidence: Literal["high", "medium", "low"] = "medium"
    purchase_url: str | None = None
    bookability: Literal["search_link", "verified_offer", "reference_only"] = "search_link"
    rank: int | None = None
    score: float | None = None
    rank_reason: str | None = None
    trip_type: Literal["one_way", "round_trip"] = "one_way"
    return_leg: FlightLegSnapshot | None = None
    flight_numbers: list[str] = Field(default_factory=list)


class FlightSearchResponse(BaseModel):
    request: FlightSearchRequest
    fetched_at: str
    purchase: PurchaseChannel
    offers: list[FlightQuote] = Field(default_factory=list)
    ranked: list[FlightQuote] = Field(default_factory=list)
    sources_used: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    errors: dict[str, str] = Field(default_factory=dict)
    latency_ms: dict[str, int] = Field(default_factory=dict)


class FlightIntentMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(..., min_length=1, max_length=4000)


class FlightVerifyFromTextRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=2000)
    chat_history: list[FlightIntentMessage] = Field(default_factory=list, max_length=20)


class FlightIntentParsed(BaseModel):
    origin: str | None = None
    destination: str | None = None
    date: str | None = None
    return_date: str | None = None
    adults: int = Field(default=1, ge=1, le=9)
    preference: RankPreference = "balanced"
    missing_fields: list[str] = Field(default_factory=list)


class FlightVerifyFromTextResponse(BaseModel):
    reply: str
    parsed: FlightIntentParsed
    search: FlightSearchResponse | None = None
    recommendation: str | None = None
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/flight.ts
import { formatEndpointPair, formatRouteLabel } from '../data/airportLabels';

export interface FlightPurchaseChannel {
  name: 'Trip.com';
  url: string;
  type: 'search';
  note: string;
}

export type RankPreference = 'cheap' | 'fast' | 'balanced';

export type FlightTripType = 'one_way' | 'round_trip';

export type FlightSource = 'ignav' | 'letsfg' | 'manual';

export interface FlightLegSnapshot {
  origin_iata: string;
  dest_iata: string;
  airline: string;
  route_label: string;
  depart_time: string;
  arrive_time: string;
  duration_minutes: number;
  stops: number;
  flight_numbers?: string[];
}

export interface FlightSearchRequest {
  origin: string;
  destination: string;
  date: string;
  return_date?: string | null;
  adults?: number;
  cabin?: 'economy' | 'premium_economy' | 'business' | 'first';
  preference?: RankPreference;
  include_letsfg?: boolean;
  include_ignav?: boolean;
}

export interface FlightQuote {
  id: string;
  origin_iata: string;
  dest_iata: string;
  airline: string;
  route_label: string;
  depart_time: string;
  arrive_time: string;
  duration_minutes: number;
  stops: number;
  price_amount: number;
  price_currency: string;
  source: FlightSource | string;
  confidence: 'high' | 'medium' | 'low';
  rank?: number | null;
  score?: number | null;
  rank_reason?: string | null;
  bookability?: string;
  purchase_url?: string | null;
  trip_type?: FlightTripType;
  return_leg?: FlightLegSnapshot | null;
  flight_numbers?: string[];
}

export interface FlightSearchResponse {
  request: FlightSearchRequest;
  purchase: FlightPurchaseChannel;
  fetched_at: string;
  offers?: FlightQuote[];
  ranked?: FlightQuote[];
  sources_used?: string[];
  warnings?: string[];
  errors?: Record<string, string>;
  latency_ms?: Record<string, number>;
}

export interface FlightIntentMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface FlightIntentParsed {
  origin: string | null;
  destination: string | null;
  date: string | null;
  return_date: string | null;
  adults: number;
  preference: RankPreference;
  missing_fields: string[];
}

export interface FlightVerifyFromTextRequest {
  message: string;
  chat_history?: FlightIntentMessage[];
}

export interface FlightVerifyFromTextResponse {
  reply: string;
  parsed: FlightIntentParsed;
  search: FlightSearchResponse | null;
  recommendation?: string | null;
  warnings?: string[];
}

export interface FlightVerifyTurn {
  id: string;
  userMessage: string;
  response: FlightVerifyFromTextResponse;
}

const PREFERENCE_LABELS: Record<RankPreference, string> = {
  cheap: '便宜优先',
  fast: '省时优先',
  balanced: '性价比（价时中转到）',
};

export function formatPreference(pref: RankPreference): string {
  return PREFERENCE_LABELS[pref] ?? pref;
}

export function formatDuration(minutes: number): string {
  if (minutes <= 0) return '—';
  const h = Math.floor(minutes / 60);
  const m = minutes % 60;
  if (h && m) return `${h}h ${m}m`;
  if (h) return `${h}h`;
  return `${m}m`;
}

export function formatPrice(offer: FlightQuote): string {
  if (offer.price_amount <= 0) return '—';
  const suffix = offer.trip_type === 'round_trip' ? '（往返）' : '';
  return `${offer.price_currency} ${offer.price_amount.toLocaleString()}${suffix}`;
}

export function formatTimeShort(iso: string): string {
  if (!iso) return '—';
  const t = iso.includes('T') ? iso.slice(11, 16) : iso.slice(0, 5);
  return t || '—';
}

export function formatStopsLabel(stops: number): string {
  if (stops === 0) return '直飞';
  return `${stops} 停`;
}

export function formatLegRoute(
  leg: Pick<FlightLegSnapshot, 'route_label' | 'origin_iata' | 'dest_iata'>,
): string {
  if (leg.route_label?.includes('→')) {
    return formatRouteLabel(leg.route_label);
  }
  return formatEndpointPair(leg.origin_iata, leg.dest_iata);
}

export function isRoundTripQuote(offer: FlightQuote): boolean {
  return offer.trip_type === 'round_trip' && !!offer.return_leg;
}

export function totalTripDuration(offer: FlightQuote): number {
  let total = offer.duration_minutes || 0;
  if (offer.return_leg) total += offer.return_leg.duration_minutes || 0;
  return total;
}

export function formatFlightSource(source: string | undefined, bookedExternal?: boolean): string {
  if (source === 'manual') {
    return bookedExternal ? '已订·手动' : '手动录入';
  }
  if (source === 'ignav') return '参考价';
  return '';
}
